packages/devtools_app/lib/src/shared/editable_list.dart

```diff
@@ -0,0 +1,272 @@
+                entries: widget.entries,
+                onEntryRemoved: widget.onEntryRemoved,
+              ),
+            ),
+          ],
+        );
+      },
+    );
+  }
+}
+
+@visibleForTesting
+class EditableListActionBar extends StatelessWidget {
+  const EditableListActionBar({
+    Key? key,
+    required this.textFieldFocusNode,
+    required this.textFieldController,
+    required this.isRefreshing,
+    required this.textFieldLabel,
+    required this.onEntryAdded,
+    required this.onRefresh,
+  }) : super(key: key);
+
+  final FocusNode textFieldFocusNode;
+  final TextEditingController textFieldController;
+  final ValueListenable<bool>? isRefreshing;
+  final String textFieldLabel;
+  final Function(String)? onEntryAdded;
+  final Function()? onRefresh;
+
+  void _addNewItem() {
+    final value = textFieldController.value.text.trim();
+    textFieldController.clear();
+    if (onEntryAdded != null && value.isNotEmpty) {
+      textFieldController.clear();
+      onEntryAdded!(value);
+    }
+    textFieldFocusNode.requestFocus();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Container(
+      height: defaultTextFieldHeight,
+      child: Row(
+        children: [
+          Expanded(
+            child: Container(
+              height: defaultTextFieldHeight,
+              child: TextField(
+                focusNode: textFieldFocusNode,
+                controller: textFieldController,
+                decoration: InputDecoration(
+                  contentPadding: const EdgeInsets.all(denseSpacing),
+                  border: const OutlineInputBorder(),
+                  labelText: textFieldLabel,
+                ),
+                onSubmitted: (value) {
+                  _addNewItem();
+                },
+              ),
+            ),
+          ),
+          TextButton(
+            onPressed: () {
+              _addNewItem();
+            },
+            child: const Text('Add'),
+          ),
+          isRefreshing?.value ?? false
+              ? Container(
+                  width: defaultTextFieldHeight,
+                  height: defaultTextFieldHeight,
+                  child: const Padding(
+                    padding: EdgeInsets.all(densePadding),
+                    child: CircularProgressIndicator(),
+                  ),
+                )
+              : RefreshButton(
+                  onPressed: () {
+                    if (onRefresh != null) {
+                      onRefresh!();
+                    }
+                  },
+                  minScreenWidthForTextBeforeScaling: double.maxFinite,
+                ),
+        ],
+      ),
+    );
+  }
+}
+
+class _EditableListContentView extends StatelessWidget {
+  const _EditableListContentView({
+    Key? key,
+    required this.entries,
+    required this.onEntryRemoved,
+  }) : super(key: key);
+
+  final ValueListenable<List<String>> entries;
+  final Function(String)? onEntryRemoved;
```

Comment:
It seems like this should be non-nullable. If we cannot remove things from the list, that eliminates a big function that makes the list "editable"